datacube/index/_datasets.py

```diff
@@ -703,3 +703,222 @@ def search_eager(self, **query):
+
+        The datasets can be selected based on values of custom fields as well as long as relevant custom
+        offsets are provided.
+
+        :param field_names: A tuple of field names that would be returned including derived fields
+                            such as extent, crs
+        :param custom_offsets: A dictionary of offsets in the metadata doc for custom fields
+        :param limit: Number of datasets returned per product.
+        :param query: key, value mappings of query that will be processed against metadata_types,
+                      product definitions on the client side as well as dataset table.
+        :return: A Dynamically generated DatasetLight (a subclass of tuple) objects.
+        """
+
+        assert field_names
+
+        for product, query_exprs in self.make_query_expr(query, custom_offsets):
+
+            select_fields, fields_to_process = self.make_select_fields(product, field_names, custom_offsets)
+
+            result_type = namedtuple('DatasetLight', tuple(field.name for field in select_fields))
+
+            class DatasetLight(result_type):
+
+                if fields_to_process.get('grid_spatial'):
+                    @property
+                    def _gs(self):
+                        return self.grid_spatial
+
+                    @property
+                    def crs(self):
+                        return Dataset.crs.__get__(self)
+
+                    @property
+                    def extent(self):
+                        return Dataset.extent.__get__(self, Dataset)
+
+                    @property
+                    def transform(self):
+                        return Dataset.transform.__get__(self)
+
+                    @property
+                    def bounds(self):
+                        return Dataset.bounds.__get__(self)
+
+            with self._db.connect() as connection:
+                results = connection.search_unique_datasets(
+                    query_exprs,
+                    select_fields=select_fields,
+                    limit=limit
+                )
+
+            import json
+
+            for result in results:
+                field_values = {field.name: result[i_] for i_, field in enumerate(select_fields)}
+                if 'grid_spatial' in fields_to_process:
+                    field_values['grid_spatial'] = json.loads(field_values['grid_spatial'])
+                yield DatasetLight(**field_values)
+
+    def make_select_fields(self, product, field_names, custom_offsets):
+        """
+        Parse and generate the list of select fields to be passed to the database API and
+        those fields that are to be further processed once the results are returned.
+        """
+
+        assert product and field_names
+
+        dataset_fields = product.metadata_type.dataset_fields
+        dataset_section = product.metadata_type.definition['dataset']
+
+        from datacube.drivers.postgres._fields import SimpleDocField
```

Comment:
please move imports to the top of the function, or better to module level

Reply:
Done